README.md
## Z-IO

[![Hackage](https://img.shields.io/hackage/v/Z-IO.svg?style=flat)](https://hackage.haskell.org/package/Z-IO)
[![Linux Build Status](https://github.com/ZHaskell/z-io/workflows/ubuntu-ci/badge.svg)](https://github.com/ZHaskell/z-io/actions)
[![MacOS Build Status](https://github.com/ZHaskell/z-io/workflows/osx-ci/badge.svg)](https://github.com/ZHaskell/z-io/actions)
[![Windows Build Status](https://github.com/ZHaskell/z-io/workflows/win-ci/badge.svg)](https://github.com/ZHaskell/z-io/actions)
[![Docker Build Status](https://github.com/ZHaskell/z-io/workflows/docker-ci/badge.svg)](https://github.com/ZHaskell/z-io/actions)
[![Gitter chat](https://badges.gitter.im/gitterHQ/gitter.svg)](https://gitter.im/Z-Haskell/community)
<a href="https://opencollective.com/zhaskell/donate" target="_blank">
  <img src="https://opencollective.com/zhaskell/donate/button@2x.png?color=blue" width=128 />
</a>

This package is part of [Z.Haskell](https://z.haskell.world) project, provides basic IO operations:

* IO resource management, resource pool
* File system operations
* Network: DNS, TCP, UDP and IPC
* Buffered input and output
* Process management
* Environment settings
* High performance logger
* High performance low resolution timer

## Requirements

* A working haskell compiler system, GHC(>=8.6), cabal-install(>=2.4), hsc2hs.
* Tests need  [hspec-discover](https://hackage.haskell.org/package/hspec-discover).

## Example usage

```haskell
> :set -XOverloadedStrings
> import Z.IO.Network
> import Z.IO.Resource
> import Z.IO.Buffered
>
> -- call getAddrInfo to perform DNS
> head <$> getAddrInfo Nothing "www.bing.com" "http"
AddrInfo {addrFlags = [AI_ADDRCONFIG,AI_V4MAPPED], addrFamily = SocketFamily 2, addrSocketType = SocketType 1, addrProtocol = ProtocolNumber 6, addrAddress = 204.79.197.200:80, addrCanonName = ""}
>
> import qualified Z.Data.Text as T
> -- send a simple HTTP request
> :{
let addr = ipv4 "13.107.21.200" 80
in withResource (initTCPClient defaultTCPClientConfig{ tcpRemoteAddr = addr}) $ \ tcp -> do
    (i, o) <- newBufferedIO tcp
    writeBuffer' o "GET http://www.bing.com HTTP/1.1\r\nHost: www.bing.com\r\n\r\n"
    readBuffer i >>= pure . T.validate
:}
"HTTP/1.1 200 OK\r\nDate: Sat, 19 Sep 2020 06:11:08 GMT\r\nContent-Length: 0\r\n\r\n"
>
> -- Start a TCP echo server, use @nc -v localhost 8080@ to test
> :{
startTCPServer defaultTCPServerConfig{
    tcpListenAddr = SocketAddrIPv4 ipv4Loopback 8080} $ \ tcp -> do
        i <- newBufferedInput tcp
        o <- newBufferedOutput tcp
        forever $ readBuffer i >>= writeBuffer o >> flushBuffer o
}
:}
```

## Dev guide

```bash
# get code
git clone --recursive git@github.com:ZHaskell/z-io.git
cd z-io
# build
cabal build
# test
cabal run Z-IO-Test
# install
cabal install
# generate document
cabal haddock
```


ChangeLog.md
# Revision history for Z-IO

## 1.0.1.0  -- 2020-07-24

Maintenance release:

* Update libuv to v1.42.0, rewrite `UV_Stream`'s `Output` instance to perform a `uv_try_write` first since now `uv_try_write` is sensible.

## 1.0.0.0  -- 2020-07-08

* Clean function names in `Z.IO.BIO` module, now no `BIO` or `Node` suffix anymore.
* `Z.IO.BIO` is not re-exported from `Z.IO` anymore, user are recommended to import it with qualified name, e.g. `import qualified Z.IO.BIO as BIO`.
* Add `foldl'` and `foldIO'` to `Z.IO.BIO` to use with `Fold` and `FoldM` from `foldl` package.
* Add `INLINABLE` pragmas to many functions.
* Add `printStdLnP` to `Z.IO.StdStream`, a `Parser` debug tool.

## 0.8.1.0  -- 2020-06-12

* Remove `-march=native` flag to improve binary portability.

## 0.8.1.0  -- 2020-04-25

* Add `getInterface` to `Z.IO.Network`.
* `mkstemp` now return opend file, the type changed to `mkstemp :: CBytes -> CBytes -> Bool -> Resource (CBytes, File)`, which has an option for keep file or not.
* `initTempFile` and `initTempDir` now do not need a prefix argument, the prefix is hardcoded as `Z-IO-`.

## 0.8.0.0  -- 2020-04-25

This is an experimental version to test new 'BIO' module.

* Rewrite `Z.IO.BIO` module, now `BIO` is push based.
* Remove `>|>`, `>~>`, `>!>`, now `BIO` nodes can be composed via funtion composition `(.)`!
* Remove `zipSource/zipBIO`, add `stepBIO/stepBIO_/runBIO_`.
* Add `zipBIO` to `Z.IO.BIO.Concurrent`, which run two BIO nodes concurrently.
* Add `ungroupingNode`, change `newGroupingNode` to use `Vector`.
* Rename `EOF` exception to `UnexpectedEOF` to avoid the clash with `EOF` pattern.

## 0.7.1.0  -- 2020-03-16

* Use `CPtr` from Z-Data instead of `ForeignPtr`.

## 0.7.0.0  -- 2020-03-09

* Change resource `Pool` to keyed by default, add `SimplePool`.
* Add `Semigroup` instance to `Logger`.
* Add `clearInputBuffer/clearOutputBuffer` to `Z.IO.Buffered`.
* Add `catchSync/ingoreSync` to `Z.IO.Exception`.
* Add `putStdLn/printStdLn` back.

## 0.6.4.0  -- 2020-02-20

* Add `initProcess'` to kill process while finish using the process resource by default.

## 0.6.3.0  -- 2020-02-20

* Split `Z.IO.UV.FFI` to `Z.IO.UV.FFI` and `Z.IO.UV.FFI_Env`, to make the module buildable when memory is constrained.
* Make functions works on TTY in `Z.IO.StdStream` correctly ignore redirected streams.
* Move `pathSeparator` to `pathSeparators`, now `pathSeparator` return the default path separator.

## 0.6.2.0  -- 2020-02-18

* Hide `Logger` constructor from `Z.IO.Logger`, remove implementation details such as `defaultTSCache`, `pushLogIORef`, `flushLogIORef`, add `loggerFormatter` to `LoggerConfig`.
* Add `newStdLogger/newFileLogger` to make new logger easily.
* Rework `Z.IO.FileSystem.Watch`'s API, change `watchDirs` to accept a recursive param and a callback.
* Hide `Z.IO.UV.Win` module, which should not be directly used by user.
* Fix a bug when stdio is redirected to pipes: https://github.com/ZHaskell/z-io/pull/16

## 0.6.1.0  -- 2020-02-09

* Fix a bug in `newMagicSplitter/newLineSplitter` code.
* Remove `sourceFromInput` and related functions to reduce API surface, use `newBufferedInput` with `sourceFromBuffered` instead.
* Refactor server loop to allow more code sharing between `Z.IO.Network.TCP` and `Z.IO.Network.IPC`.

## 0.6.0.0  -- 2020-02-04

* FileSystem: replace `DEFAULT_MODE` with `DEFAULT_FILE_MODE` & `DEFAULT_DIR_MODE`.
* Ignore exception while `mkdirp` on an exist directory.
* Make `rmrf` more like `rm -rf`, which can be used on files.
* Add `doesPathExist/doesFileExist/doesDirExist` to file system module.
* Add `Z.IO.FileSystem` re-export `Z.IO.FileSystem.Watch` and `Z.IO.FileSystem.FilePath`.
* Add `mkstemp`, `initTempFile/initTempDir` to file system module.

## 0.5.0.0  -- 2020-01-28

* Add `unwrap/unwrap'` to `Z.IO.Exception`.
* Add `readParseChunks` to `Z.IO.Buffered`, Change `readParser`'s type to match `readParseChunks`.
* Add `sourceParseChunksBufferedInput`, `sourceParseChunksInput` to `Z.IO.BIO`.
* Add `newJSONLogger/defaultJSONFmt` to `Z.IO.Logger`, provide simple JSON structured logging.

## 0.3.0.0  -- 2020-12-29

* Add `getSystemTime'` to `Z.IO.Time`.
* Add `shutdownUVStream` to `Z.IO.UV.UVStream`.
* Change `sourceFrom/sinkToFile` to `initSourceFrom/initSinkToFile`.
* Bump `Z-Data` version.

## 0.2.0.0  -- 2020-12-16

* Add `sourceParsedBufferInput` and JSON sources to `Z.IO.BIO`.
* Fix `readLine` and `newLineSplitter`.
* Improve low resolution timer precision.
* Fix a bug in `Z.IO.FileSystem.FilePath.relative`, see [#17](https://github.com/likle/cwalk/issues/17).

## 0.1.9.0  -- 2020-11-23

* Clean up API in `Z.IO.Buffered`, remove `readToMagic'`, `readLine'`, `readExactly'`.
* `readExactly` now throw exception when not reading enough bytes before EOF.
* Add `Show/ShowT` instance to `UVStream`, `StdStream`, `UDP`, `UVManager`.
* Add JSON instance to various types: `SocketAddr` and all configure types.
* Rename `InetAddr` to `IPv4`, `Inet6Addr` to `IPv6`, change `SocketAddr` 's constructor name, and payload order.
* Add `seek` to `Z.IO.FileSystem`.

## 0.1.8.1  -- 2020-11-21

* Export `ZStream` type from `Z.IO.BIO.Zlib`

## 0.1.8.0  -- 2020-11-20

* Remove type index from `BufferedInput`, `BufferedOutput`.
* Add `Z.IO.BIO` module to facilitate streaming process, and `Z.IO.BIO.Concurrent` to facilitate producer-consumer model.
* Remove streamming related functions from `Z.IO.Buffered`, use `Z.IO.BIO` instead.
* Move `Z.IO.Compression.Zlib` to `Z.IO.BIO.Zlib`, change API to `BIO` style.
* Add `Z.IO.FileSystem.Watch` module, provides cross-platform filesystem watching.

## 0.1.7.0  -- 2020-10-24

* Add `iso8016DateFormat`, change logger's default time format to include time zone.
* Rename `warn` to `warning`, change `Level` to `Int` type alias in `Z.IO.Logger`, add `critical`.
* Export `TimeVal` from `Z.IO.Environment`.
* Add `getCPUInfo`, `getLoadAvg`, `getXXXMem` to `Z.IO.Environment`.

## 0.1.6.1  -- 2020-10-17

* Export `ResUsage` from `Z.IO.Environment`.

Z/IO/Logger.hs
{-# LANGUAGE LambdaCase #-}

{-|
Module      : Z.IO.Logger
Description : High performance logger
Copyright   : (c) Dong Han, 2017-2018
License     : BSD
Maintainer  : winterland1989@gmail.com
Stability   : experimental
Portability : non-portable

Simple, high performance logger. The design choice of this logger is biased towards simplicity instead of generlization:

* All log functions live in 'IO'.
* By default this logger is connected to stderr, use 'setDefaultLogger' to customize.
* When logging each thread will build log 'Builder's into a small 'V.Bytes' with line buffer instead of leaving all 'Builder's to the flushing thread:

    * Logger won't keep heap data for too long simply because they're referenced by log's 'Builder'.
    * Each logging thread only need perform a CAS to prepend log 'V.Bytes' into a list, which reduces contention.
    * Each log call is atomic, Logging order is preserved under concurrent settings.

Flushing is automatic and throttled for 'debug', 'info', 'warning' to boost
performance, but a 'fatal' and 'critical' log will always flush logger's buffer.
This could lead to a problem that if main thread exits too early logs may missed,
to add a flushing when program exits, use 'withDefaultLogger' like:

@
import Z.IO.Logger

main :: IO ()
main = withDefaultLogger $ do
    ....
    debug "..."   -- So that this log won't be missed
    ...
@
-}

module Z.IO.Logger
  ( -- * A simple Logger type
    Logger
  , LoggerConfig(..)
  , defaultLoggerConfig
  , defaultJSONLoggerConfig
  , setDefaultLogger
  , getDefaultLogger
  , flushDefaultLogger
  , withDefaultLogger

    -- * Create a new logger
  , newLogger
  , newStdLogger
  , newFileLogger

    -- * logging functions
  , debug
  , info
  , warning
  , fatal
  , critical
  , otherLevel

    -- * logging functions with specific logger
  , debugTo
  , infoTo
  , warningTo
  , fatalTo
  , otherLevelTo

    -- * Helpers to write new log formatter
  , LogFormatter, defaultFmt, defaultColoredFmt, defaultJSONFmt
  , defaultFmtCallStack
  , defaultLevelFmt

    -- * Constants
    -- ** Level
  , Level
  , pattern DEBUG
  , pattern INFO
  , pattern WARNING
  , pattern FATAL
  , pattern CRITICAL
  , pattern NOTSET
  ) where

import           Control.Concurrent.MVar
import           Control.Monad
import           Data.Bits               ((.|.))
import           Data.IORef
import           Foreign.C.Types         (CInt (..))
import           GHC.Conc.Sync           (ThreadId (..), myThreadId)
import           GHC.Exts                (ThreadId#)
import           GHC.Stack
import           System.IO.Unsafe        (unsafePerformIO)
import qualified Z.Data.Builder          as B
import qualified Z.Data.CBytes           as CB
import qualified Z.Data.JSON.Builder     as JB
import qualified Z.Data.Vector.Base      as V
import           Z.IO.Buffered
import           Z.IO.Exception
import qualified Z.IO.FileSystem         as ZF
import           Z.IO.LowResTimer
import           Z.IO.Resource
import           Z.IO.StdStream
import           Z.IO.StdStream.Ansi     (AnsiColor (..), color)
import           Z.IO.Time

-------------------------------------------------------------------------------

-- | Formatter used by `Logger`.
type LogFormatter = B.Builder ()            -- ^ data\/time string(second precision)
                  -> Level                  -- ^ log level
                  -> B.Builder ()           -- ^ log content
                  -> CallStack              -- ^ call stack trace
                  -> ThreadId               -- ^ logging thread id
                  -> B.Builder ()

-- | The `Logger` type.
data Logger = Logger
    (Level -> Bool ->  CallStack -> B.Builder () -> IO ())  -- ^ logging function
    (IO ())                                                 -- ^ manually flush

-- | For composing different loggers
instance Semigroup Logger where
    Logger log1 flush1 <> Logger log2 flush2 = Logger
        (\ l b cs bu -> log1 l b cs bu >> log2 l b cs bu)
        (flush1 >> flush2)

-- | Logger config type used in this module.
data LoggerConfig = LoggerConfig
    { loggerMinFlushInterval :: {-# UNPACK #-} !Int
    -- ^ Minimal flush interval, see Notes on 'debug'
    , loggerLineBufSize      :: {-# UNPACK #-} !Int
    -- ^ Buffer size to build each log line
    , loggerLevel            :: {-# UNPACK #-} !Level
    -- ^ Config log's filter level
    , loggerFormatter        :: LogFormatter
    -- ^ Log formatter
    }

-- | A default logger config with
--
-- * 0.1s minimal flush interval
-- * line buffer size 240 bytes
-- * show everything by default
-- * 'defaultFmt'
defaultLoggerConfig :: LoggerConfig
{-# INLINABLE defaultLoggerConfig #-}
defaultLoggerConfig = LoggerConfig 1 240 NOTSET defaultFmt

-- | A default logger config with
--
-- * 0.5s minimal flush interval
-- * line buffer size 1000 bytes
-- * show everything by default
-- * 'defaultJSONFmt'
defaultJSONLoggerConfig :: LoggerConfig
{-# INLINABLE defaultJSONLoggerConfig #-}
defaultJSONLoggerConfig = LoggerConfig 5 1000 NOTSET defaultJSONFmt

-- | A default timestamp cache with format @%Y-%m-%dT%H:%M:%S%Z@('iso8061DateFormat').
--
-- The timestamp will updated in 0.1s granularity to ensure a seconds level precision.
defaultTSCache :: IO (B.Builder ())
{-# NOINLINE defaultTSCache #-}
defaultTSCache = unsafePerformIO $ do
    throttle 1 $ do
        t <- getSystemTime'
        CB.toBuilder <$> formatSystemTime iso8061DateFormat t

-------------------------------------------------------------------------------

-- | Make a new logger with given write device.
newLogger :: LoggerConfig
          -> MVar BufferedOutput
          -> IO Logger
{-# INLINABLE newLogger #-}
newLogger LoggerConfig{..} oLock = do
    logsRef <- newIORef []
    let flush = flushLogIORef oLock logsRef
    throttledFlush <- throttleTrailing_ loggerMinFlushInterval flush
    return $ Logger (\ level flushNow cstack bu ->
            when (level >= loggerLevel) $ do
                ts <- defaultTSCache
                tid <- myThreadId
                (pushLogIORef logsRef loggerLineBufSize) $ loggerFormatter ts level bu cstack tid
                if flushNow then flush else throttledFlush
        ) flush


-- | Make a new logger write to 'stderrBuf'.
newStdLogger :: LoggerConfig -> IO Logger
{-# INLINABLE newStdLogger #-}
newStdLogger config = newLogger config stderrBuf

-- | Make a new file based logger with 'defaultFmt'.
--
-- The file will be opened in append mode.
newFileLogger :: LoggerConfig -> CB.CBytes -> IO Logger
{-# INLINABLE newFileLogger #-}
newFileLogger config path = do
    let res = ZF.initFile path (ZF.O_CREAT .|. ZF.O_RDWR .|. ZF.O_APPEND) ZF.DEFAULT_FILE_MODE
    (file, _closeFunc) <- acquire res
    oLock <- newMVar =<< newBufferedOutput file
    newLogger config oLock

-------------------------------------------------------------------------------

-- | Use 'pushLogIORef' and 'pushLogIORef' to implement a simple 'IORef' based concurrent logger.
--
-- @
-- logsRef <- newIORef []
-- let push = pushLogIORef logsRef lineBufSize
--     flush = flushLogIORef stderrBuf logsRef
--     throttledFlush <- throttleTrailing_ flushInterval flush
-- ..
-- return $ Logger push flush throttledFlush ...
-- @
--
pushLogIORef :: IORef [V.Bytes]     -- ^ logs stored in a list, new log will be CASed into it.
             -> Int                 -- ^ buffer size to build each log
             -> B.Builder ()        -- ^ formatted log
             -> IO ()
{-# INLINABLE pushLogIORef #-}
pushLogIORef logsRef loggerLineBufSize b = do
    let !bs = B.buildWith loggerLineBufSize b
    unless (V.null bs) $ atomicModifyIORef' logsRef (\ bss -> (bs:bss, ()))

flushLogIORef :: HasCallStack => MVar BufferedOutput -> IORef [V.Bytes] -> IO ()
{-# INLINABLE flushLogIORef #-}
flushLogIORef oLock logsRef =
    withMVar oLock $ \ o -> do
        bss <- atomicModifyIORef' logsRef (\ bss -> ([], bss))
        forM_ (reverse bss) (writeBuffer o)
        flushBuffer o

-- | A default log formatter
--
-- @[FATAL][2021-02-01T15:03:30+0800][<interactive>:31:1][thread#669]...@
defaultFmt :: LogFormatter
{-# INLINABLE defaultFmt #-}
defaultFmt ts level content cstack (ThreadId tid#) = do
    B.square (defaultLevelFmt level)
    B.square ts
    B.square $ defaultFmtCallStack cstack
    B.square $ "thread#" >> B.int (getThreadId tid#)
    content
    B.char8 '\n'

-- | A default colored log formatter
--
-- DEBUG level is 'Cyan', WARNING level is 'Yellow', FATAL and CRITICAL level are 'Red'.
defaultColoredFmt :: LogFormatter
{-# INLINABLE defaultColoredFmt #-}
defaultColoredFmt ts level content cstack (ThreadId tid#) = do
    let blevel = defaultLevelFmt level
    B.square (case level of
        DEBUG    -> color Cyan blevel
        WARNING  -> color Yellow blevel
        FATAL    -> color Red blevel
        CRITICAL -> color Red blevel
        _        -> blevel)
    B.square ts
    B.square $ defaultFmtCallStack cstack
    B.square $ "thread#" >> B.int (getThreadId tid#)
    content
    B.char8 '\n'

-- | A default JSON log formatter.
--
-- > {"level":"FATAL","time":"2021-02-01T15:02:19+0800","loc":"<interactive>:27:1","theadId":606,"content":"..."}\n
defaultJSONFmt :: LogFormatter
{-# INLINABLE defaultJSONFmt #-}
defaultJSONFmt ts level content cstack (ThreadId tid#) = do
    B.curly $ do
        "level" `JB.kv` B.quotes (defaultLevelFmt level)
        B.comma
        "time" `JB.kv` B.quotes ts
        B.comma
        "loc" `JB.kv` B.quotes (defaultFmtCallStack cstack)
        B.comma
        "thead" `JB.kv`  B.int (getThreadId tid#)
        B.comma
        "content" `JB.kv` JB.string (B.unsafeBuildText content)
    B.char8 '\n'

-- | Default stack formatter which fetch the logging source and location.
defaultFmtCallStack :: CallStack -> B.Builder ()
{-# INLINABLE defaultFmtCallStack #-}
defaultFmtCallStack cs =
 case reverse $ getCallStack cs of
   [] -> "<no call stack found>"
   (_, loc):_ -> do
      B.string8 (srcLocFile loc)
      B.char8 ':'
      B.int (srcLocStartLine loc)
      B.char8 ':'
      B.int (srcLocStartCol loc)

globalLogger :: IORef Logger
{-# NOINLINE globalLogger #-}
globalLogger = unsafePerformIO $
    newIORef =<< newStdLogger defaultLoggerConfig{
        loggerFormatter = (if isStdStreamTTY stderr then defaultColoredFmt else defaultFmt)
    }

-- | Change the global logger.
setDefaultLogger :: Logger -> IO ()
{-# INLINABLE setDefaultLogger #-}
setDefaultLogger !logger = atomicWriteIORef globalLogger logger

-- | Get the global logger.
--
-- This is a logger connected to stderr, if stderr is connect to TTY,
-- then use 'defaultColoredFmt', otherwise use 'defaultFmt'.
getDefaultLogger :: IO Logger
{-# INLINABLE getDefaultLogger #-}
getDefaultLogger = readIORef globalLogger

-- | Manually flush global logger.
flushDefaultLogger :: IO ()
{-# INLINABLE flushDefaultLogger #-}
flushDefaultLogger = do
    (Logger _ flush) <- getDefaultLogger
    flush

-- | Flush global logger when program exits.
withDefaultLogger :: IO () -> IO ()
{-# INLINABLE withDefaultLogger #-}
withDefaultLogger = (`finally` flushDefaultLogger)

--------------------------------------------------------------------------------

-- | Logging Levels
--
-- We following the Python logging levels, for details,
-- see: <https://docs.python.org/3/howto/logging.html#logging-levels>
--
-- +----------+---------------+
-- | Level    | Numeric value |
-- +----------+---------------+
-- | CRITICAL | 50            |
-- +----------+---------------+
-- | FATAL    | 40            |
-- +----------+---------------+
-- | WARNING  | 30            |
-- +----------+---------------+
-- | INFO     | 20            |
-- +----------+---------------+
-- | DEBUG    | 10            |
-- +----------+---------------+
-- | NOTSET   | 0             |
-- +----------+---------------+
--
type Level = Int

pattern CRITICAL :: Level
pattern CRITICAL = 50

pattern FATAL :: Level
pattern FATAL = 40

pattern WARNING :: Level
pattern WARNING = 30

pattern INFO :: Level
pattern INFO = 20

pattern DEBUG :: Level
pattern DEBUG = 10

pattern NOTSET :: Level
pattern NOTSET = 0

-- | Format 'DEBUG' to 'DEBUG', etc.
--
-- Level other than built-in ones, are formatted in decimal numeric format, i.e.
-- @defaultLevelFmt 60 == "LEVEL60"@
defaultLevelFmt :: Level -> B.Builder ()
{-# INLINABLE defaultLevelFmt #-}
defaultLevelFmt level = case level of
    CRITICAL -> "CRITICAL"
    FATAL    -> "FATAL"
    WARNING  -> "WARNING"
    INFO     -> "INFO"
    DEBUG    -> "DEBUG"
    NOTSET   -> "NOTSET"
    level'   -> "LEVEL" >> B.int level'

debug :: HasCallStack => B.Builder () -> IO ()
{-# INLINABLE debug #-}
debug = otherLevel_ DEBUG False callStack

info :: HasCallStack => B.Builder () -> IO ()
{-# INLINABLE info #-}
info = otherLevel_ INFO False callStack

warning :: HasCallStack => B.Builder () -> IO ()
{-# INLINABLE warning #-}
warning = otherLevel_ WARNING False callStack

fatal :: HasCallStack => B.Builder () -> IO ()
{-# INLINABLE fatal #-}
fatal = otherLevel_ FATAL True callStack

critical :: HasCallStack => B.Builder () -> IO ()
{-# INLINABLE critical #-}
critical = otherLevel_ CRITICAL True callStack

otherLevel :: HasCallStack
           => Level             -- ^ log level
           -> Bool              -- ^ flush immediately?
           -> B.Builder ()      -- ^ log content
           -> IO ()
{-# INLINABLE otherLevel #-}
otherLevel level flushNow bu = otherLevel_ level flushNow callStack bu

otherLevel_ :: Level -> Bool -> CallStack -> B.Builder () -> IO ()
{-# INLINABLE otherLevel_ #-}
otherLevel_ level flushNow cstack bu = do
    (Logger f _) <- getDefaultLogger
    f level flushNow cstack bu

--------------------------------------------------------------------------------

debugTo :: HasCallStack => Logger -> B.Builder () -> IO ()
{-# INLINABLE debugTo #-}
debugTo = otherLevelTo_ DEBUG False callStack

infoTo :: HasCallStack => Logger -> B.Builder () -> IO ()
{-# INLINABLE infoTo #-}
infoTo = otherLevelTo_ INFO False callStack

warningTo :: HasCallStack => Logger -> B.Builder () -> IO ()
{-# INLINABLE warningTo #-}
warningTo = otherLevelTo_ WARNING False callStack

fatalTo :: HasCallStack => Logger -> B.Builder () -> IO ()
{-# INLINABLE fatalTo #-}
fatalTo = otherLevelTo_ FATAL True callStack

otherLevelTo :: HasCallStack
             => Logger
             -> Level             -- ^ log level
             -> Bool              -- ^ flush immediately?
             -> B.Builder ()      -- ^ log content
             -> IO ()
{-# INLINABLE otherLevelTo #-}
otherLevelTo logger level flushNow = otherLevelTo_ level flushNow callStack logger

otherLevelTo_ :: Level -> Bool -> CallStack -> Logger -> B.Builder () -> IO ()
{-# INLINABLE otherLevelTo_ #-}
otherLevelTo_ level flushNow cs (Logger f _) = f level flushNow cs

foreign import ccall unsafe "rts_getThreadId" getThreadId :: ThreadId# -> CInt
